pkg/controller/node/node_controller.go

```diff
@@ -128,11 +154,197 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
+	oldScheduler := old.(*configv1.Scheduler)
+	curScheduler := cur.(*configv1.Scheduler)
+
+	if oldScheduler.Name != schedulerCRName || curScheduler.Name != schedulerCRName {
+		glog.V(4).Infof("We don't care about CRs other than cluster created for scheduler config")
+		return
+	}
+
+	if reflect.DeepEqual(oldScheduler.Spec, curScheduler.Spec) {
+		glog.V(4).Info("Scheduler config did not change")
+		return
+	}
+	// check the flag in scheduler spec
+	areMastersSchedulable := curScheduler.Spec.MastersSchedulable
+	currentMasters, err := ctrl.getCurrentMasters()
+	if err != nil {
+		goerrs.Wrap(err, "Reconciling to make master nodes schedulable/unschedulable failed")
+		return
+	}
+
+	// reconcile master labels and taints to make the master node schedulable, unschedulable
+	err = ctrl.reconcileSchedulableMasters(areMastersSchedulable, currentMasters)
+	if err != nil {
+		goerrs.Wrap(err, "Reconciling to make master nodes schedulable/unschedulable failed")
+		return
+	}
+	return
+}
+
+// reconcileSchedulableMasters reconciles the master nodes to make them schedulable or unschedulable
+func (ctrl *Controller) reconcileSchedulableMasters(areMastersSchedulable bool, currentMasters []*corev1.Node) error {
+	var errs []error
+	if areMastersSchedulable {
+		errs = ctrl.makeMastersSchedulable(currentMasters)
+	} else {
+		errs = ctrl.makeMastersUnSchedulable(currentMasters)
+	}
+	return v1helpers.NewMultiLineAggregate(errs)
+}
+
+// makeMastersSchedulable makes all the masters in the cluster schedulable
+func (ctrl *Controller) makeMastersSchedulable(currentMasters []*corev1.Node) []error {
+	var errs []error
+	for _, node := range currentMasters {
+		if CheckMasterIsAlreadySchedulable(node) {
+			continue
+		}
+		if err := ctrl.makeMasterNodeSchedulable(node); err != nil {
+			errs = append(errs, fmt.Errorf("failed making node %v schedulable with error %v", node.Name, err))
+		}
+	}
+	return errs
+}
+
+// makeMastersUnSchedulable makes all the masters in the cluster unschedulable
+func (ctrl *Controller) makeMastersUnSchedulable(currentMasters []*corev1.Node) []error {
+	var errs []error
+	for _, node := range currentMasters {
+		if !CheckMasterIsAlreadySchedulable(node) {
+			continue
+		}
+		if err := ctrl.makeMasterNodeUnSchedulable(node); err != nil {
+			errs = append(errs, fmt.Errorf("failed making node %v schedulable with error %v", node.Name, err))
+		}
+	}
+	return errs
+
+}
+
+// makeMasterNodeUnSchedulable makes master node unschedulable by removing worker label and adding `NoSchedule`
+// master taint to the master node
+func (ctrl *Controller) makeMasterNodeUnSchedulable(node *corev1.Node) error {
+	_, err := internal.UpdateNodeRetry(ctrl.kubeClient.CoreV1().Nodes(), ctrl.nodeLister, node.Name, func(node *corev1.Node) {
+		// Remove worker label
+		newLabels := node.Labels
+		if _, hasWorkerLabel := newLabels[workerLabel]; hasWorkerLabel {
+			delete(newLabels, workerLabel)
+		}
+		node.Labels = newLabels
```

Comment:
I  think all of the code above can just be:
`delete(node.Labels, workerLabel)` since it's a no-op if the key doesn't exist.